### PR TITLE
JAMES-3009 ADR for using scala in the event sourcing modules

### PR DESCRIPTION
**Pull request by remk**

### CODE REVIEW

File: src/adr/0016-use-scala-in-event-sourcing-modules.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+Date: 2019-12-13
+
+## Status
+
+Proposed
+
+## Context
+
+At the time being James use the scala programming language in some parts of its code base, particularily for implementing the Distributed Task Manager,
+which use the event sourcing modules.
```
Comment (Arsnael) [inline src/adr/0016-use-scala-in-event-sourcing-modules.md]: s/use/uses

File: src/adr/0016-use-scala-in-event-sourcing-modules.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+The module `event-store-memory` already use scala.
```
Comment (Arsnael) [inline src/adr/0016-use-scala-in-event-sourcing-modules.md]: s/use/uses

File: src/adr/0016-use-scala-in-event-sourcing-modules.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  -  `event-sourcing-core`
+  -  `event-sourcing-pojo`
+  -  `event-store-api`
+  -  `event-store-cassandra`
+
+
+## Consequences
+
+This will help to standardize the `event-*` modules as `event-store-memory` is already written in scala.
+This change will avoid interopability concerns with the mains consumers of those modules which are already written in scala : see the distributed task manager.
```
Comment (Arsnael) [inline src/adr/0016-use-scala-in-event-sourcing-modules.md]: s/mains/main

**Comment by mbaechler** [comment]
The construction of our ADRs looks weird as it doesn't expose alternatives and rationales.
I don't know why I didn't notice earlier but here the rationales are in Consequences and it doesn't make much sense.
Could we agree on a better ADR format?

**Comment by chibenwa** [comment]
I tend to put rationals in the context part to be honnest, but having it as a separated section would be a win IMO.

Alternatives have to be mentioned if we did actually take the time to evaluate them (and deserves a section also after *consequences*) - but this should not necessarily be compulsory. I tended to put them in reference (as that was a link to a POC that was not adopted).

**Comment by MichaelBailly** [comment]
Hi devs,

I agree with Matthieu & Benoit, that alternatives should be exposed in this ADR. Furthermore, the thing saying "hey, we got 1 module in scala and all the others in Java, so we'll recode all the others in scala, and also by the way, we have to implement facades" does not seem a natural conclusion...

IMHO, there is a need for a more project-wide ADR, than just this one. For example:

- an ADR exploring the different languages running on the JVM, or some of them, that would benefit to James, and choosing one, and saying that from now on, the new developments will happen on that techno
- or the same thing scoped to a sub-group of the James project
- or a ADR saying: ok, we went all-in in scala to test a bit, and yeah we like it because X and Y, so we decide to broaden the use of Scala to the whole project/this subset

TL:DR : I think this kind of change deserve a more global vision of the project and where in the project we want to use it

Don't be confused, I'm not against Scala, in fact I'm in favour of it, for recruitment & "more modern and less verbose" reasons.

**Comment by mbaechler** [comment]
> Hi devs,
> 
> I agree with Matthieu & Benoit, that alternatives should be exposed in this ADR. Furthermore, the thing saying "hey, we got 1 module in scala and all the others in Java, so we'll recode all the others in scala, and also by the way, we have to implement facades" does not seem a natural conclusion...
> 
> IMHO, there is a need for a more project-wide ADR, than just this one. For example:
> 
>     * an ADR exploring the different languages running on the JVM, or some of them, that would benefit to James, and choosing one, and saying that from now on, the new developments will happen on that techno

We did. The two candidates were Kotlin and Scala. If you know of credible alternatives, let me know.

However, we did years ago (ok, like 1.5 or 2 years) and Rémi don't really know about that because he joined us after that decision (and by the way, he joined us partly because of that choice).

TL;DR: Kotlin and its ecosystem is not good enough for us given that we focus more on functional programming than on "pragamtism" (it's the official selling point of Kotlin).

>     * or the same thing scoped to a sub-group of the James project

The choice made is to adopt it progressively on the entire codebase but first in places where the ROI is the higher.

The CQRS/ES codebase does benefits from that because ADT (Algebraic data type) and Pattern Matching is very important in this architecture.

Note that we write ADR for only some months and it's quite hard to back-document years of discussions and decision.

>     * or a ADR saying: ok, we went all-in in scala to test a bit, and yeah we like it because X and Y, so we decide to broaden the use of Scala to the whole project/this subset
>
> 
> TL:DR : I think this kind of change deserve a more global vision of the project and where in the project we want to use it
> 
> Don't be confused, I'm not against Scala, in fact I'm in favour of it, for recruitment & "more modern and less verbose" reasons.

Thank you for your support. We can document what we like about Scala and our long-term strategy in one or several ADRs and your comments are welcome. However, don't overlook that we do actually make a documentation effort and sometime, aiming for too much just has the opposite effect than the one we wish: it can hinder the effort.

I'm personally ok to merge this ADR and try to contribute other ones in the futur.

**Comment by remk** [comment]
Just incremented the number of the ADR.

File: src/adr/0031-use-scala-in-event-sourcing-modules.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# 31. Use scala in event sourcing modules
+
+Date: 2019-12-13
+
+## Status
+
+Proposed
```
Comment (rouazana) [inline src/adr/0031-use-scala-in-event-sourcing-modules.md:7]: should be `Accepted (lazy consensus)` now?

File: src/adr/0031-use-scala-in-event-sourcing-modules.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+## Context
+
+At the time being James use the scala programming language in some parts of its code base, particularily for implementing the Distributed Task Manager,
+which uses the event sourcing modules.
+
+The module `event-store-memory` already uses scala.
```
Comment (rouazana) [inline src/adr/0031-use-scala-in-event-sourcing-modules.md:14]:
```suggestion
The module `event-store-memory` already uses Scala.
```

Comment (rouazana) [reply src/adr/0031-use-scala-in-event-sourcing-modules.md:14]: a reference to adr #0024-polyglot-strategy should be added

File: src/adr/0031-use-scala-in-event-sourcing-modules.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+## Decision
+
+What is proposed here, is to convert in scala the event sourcing modules.
```
Comment (rouazana) [inline src/adr/0031-use-scala-in-event-sourcing-modules.md:18]:
```suggestion
What is proposed here, is to convert in Scala the event sourcing modules.
```

File: src/adr/0031-use-scala-in-event-sourcing-modules.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+The modules concerned by this change are :
```
Comment (rouazana) [inline src/adr/0031-use-scala-in-event-sourcing-modules.md:19]:
```suggestion
The modules concerned by this change are:
```

File: src/adr/0031-use-scala-in-event-sourcing-modules.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+What is proposed here, is to convert in scala the event sourcing modules.
+The modules concerned by this change are : 
+  -  `event-sourcing-core`
+  -  `event-sourcing-pojo`
+  -  `event-store-api`
+  -  `event-store-cassandra`
+
+## Rationales
+
+This will help to standardize the `event-*` modules as `event-store-memory` is already written in scala.
```
Comment (rouazana) [inline src/adr/0031-use-scala-in-event-sourcing-modules.md:27]:
```suggestion
This will help to standardize the `event-*` modules as `event-store-memory` is already written in Scala.
```

File: src/adr/0031-use-scala-in-event-sourcing-modules.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+This change will avoid interopability concerns with the main consumers of those modules which are already written in scala : see the distributed task manager.
```
Comment (rouazana) [inline src/adr/0031-use-scala-in-event-sourcing-modules.md:28]:
```suggestion
This change will avoid interopability concerns with the main consumers of those modules which are already written in Scala: see the distributed task manager.
```

File: src/adr/0031-use-scala-in-event-sourcing-modules.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+In the long run this will allow to have a stronger typing in those parts of the code and to have a much less verbose code.
+
+
+## Consequences
+
+We will have to mitigate the pervading of the scala api in the java code base by implementing java facade.
```
Comment (rouazana) [inline src/adr/0031-use-scala-in-event-sourcing-modules.md:34]:
```suggestion
We will have to mitigate the pervading of the Scale API in the Java code base by implementing Java facade.
```

**Comment by chibenwa** [comment]
https://github.com/chibenwa/james-project/commit/5d34a28d7888d90f3cf5372ec0f88a3420ed6095 should help you fixing my latest remarks.

You can cherry-pick it here.

**Comment by remk** [comment]
> [chibenwa@5d34a28](https://github.com/chibenwa/james-project/commit/5d34a28d7888d90f3cf5372ec0f88a3420ed6095) should help you fixing my latest remarks.
> 
> You can cherry-pick it here.

Wrong PR ?

**Comment by chibenwa** [comment]
Merged
